src/Data/Singletons/Single/Type.hs

```diff
@@ -32,25 +33,48 @@ singType :: OSet Name      -- the set of bound kind variables in this scope
                 , DKind )  -- the kind of the result type
 singType bound_kvs prom ty = do
   checkVanillaDType ty
-  let (_, cxt, args, res) = unravelVanillaDType ty
-      num_args            = length args
+  let (orig_tvbs, cxt, args, res) = unravelVanillaDType ty
+      num_args                    = length args
   cxt' <- mapM singPred_NC cxt
   arg_names <- replicateM num_args (qNewName "t")
   prom_args <- mapM promoteType_NC args
   prom_res  <- promoteType_NC res
   let args' = map (\n -> singFamily `DAppT` (DVarT n)) arg_names
       res'  = singFamily `DAppT` (foldl apply prom (map DVarT arg_names) `DSigT` prom_res)
       tau   = ravel args' res'
-      -- Make sure to subtract out the bound variables currently in scope, lest we
-      -- accidentally shadow them in this type signature.
-      kv_names_to_bind = foldMap fvDType (prom_args ++ cxt' ++ [prom_res])
-                            OSet.\\ bound_kvs
-      kvs_to_bind      = toList kv_names_to_bind
-  let ty' = DForallT ForallInvis
-                     (map DPlainTV kvs_to_bind ++ zipWith DKindedTV arg_names prom_args) $
-            DConstrainedT cxt' tau
+      kvbs  = singTypeKVBs orig_tvbs prom_args cxt' prom_res
+                           (map DPlainTV (toList bound_kvs))
+      all_tvbs = kvbs ++ zipWith DKindedTV arg_names prom_args
+      ty' = DForallT ForallInvis all_tvbs $ DConstrainedT cxt' tau
   return (ty', num_args, arg_names, cxt, prom_args, prom_res)
 
+-- Compute the kind variable binders to use in the singled version of a type
+-- signature. This has two main call sites: singType and D.S.Single.Data.singCtor.
+--
+-- This implements the advice documented in
+-- Note [Preserve the order of type variables during singling], wrinkle 1.
+singTypeKVBs ::
+     [DTyVarBndr] -- ^ The bound type variables from the original type signature.
+  -> [DType]      -- ^ The argument types of the signature (promoted).
+  -> DCxt         -- ^ The context of the signature (singled).
+  -> DType        -- ^ The result type of the signature (promoted).
+  -> [DTyVarBndr] -- ^ The type variables previously bound in the current scope.
+  -> [DTyVarBndr] -- ^ The kind variables for the singled type signature.
+singTypeKVBs orig_tvbs prom_args sing_ctxt prom_res bound_tvbs
+  | null orig_tvbs
+  -- There are no explicitly `forall`ed type variable binders, so we must
+  -- infer them ourselves.
+  = deleteFirstsBy
+      ((==) `on` extractTvbName)
+      (toposortTyVarsOf $ prom_args ++ sing_ctxt ++ [prom_res])
```

Comment:
Every time I see `toposortTyVarsOf`, one of my neurons commits harakiri. That function is just plain wrong. Yes, it's right in common scenarios, but so is much C code.

Yes, yes, I know we can't do better. I just can't help but to shudder when I see it.

Reply:
Cool.